### PR TITLE
[Bugfix] 500 Internal Server Error when tool_choice is incorrect.

### PR DESCRIPTION
**Pull request by shenoyvvarun**
## Summary
This PR fixes the KeyError when tool_choice is incorrectly provided.
```

Request:
		{
	    "model": "vllm-model",
	    "messages": [ {"role": "user", "content": "what is 5+3"}],
	    "tool_choice": {},
	    "tools": [
	        {
	                "type": "function",
	                "name": "my_caculator",
	                "description": "A simple caculator takes an expression to caculate.",
	                "parameters": {
	                    "type": "object",
	                    "properties": {
	                        "expression": {
	                            "type": "string",
	                            "description": "A expressoin to caculate."
	                        }
	                    },
	                    "required": ["expression"]
	                }
	        }
	    ]
	}
Output:
Internal Server Error(base)
```
Screenshot 2024-11-21 at 11 04 08 PM

### Checklist
- Tested using curl
- After fix:
```
{
    "object": "error",
    "message": "[{'type': 'value_error', 'loc': ('body',), 'msg': 'Value error, Incorrectly formatted `tool_choice`. Should be like `{\"type\": \"function\", \"function\": {\"name\": \"my_function\"}}`', 'input': {'model': 'vllm-model', 'messages': [{'role': 'user', 'content': 'what is 5+3'}], 'tool_choice': {}, 'tools': [{'type': 'function', 'name': 'my_caculator', 'description': 'A simple caculator takes an expression to caculate.', 'parameters': {'type': 'object', 'properties': {'expression': {'type': 'string', 'description': 'A expressoin to caculate.'}}, 'required': ['expression']}}]}, 'ctx': {'error': ValueError('Incorrectly formatted `tool_choice`. Should be like `{\"type\": \"function\", \"function\": {\"name\": \"my_function\"}}`')}}]",
    "type": "BadRequestError",
    "param": null,
    "code": 400
}
```
- Lint
- pytest

### CODE REVIEW

File: vllm/entrypoints/openai/protocol.py [hunk 1]
```diff
@@ -478,13 +478,13 @@ def check_tool_usage(cls, data):
             # it matches a valid tool
             if isinstance(data["tool_choice"], dict):
                 valid_tool = False
-                specified_function = data["tool_choice"]["function"]
+                specified_function = data["tool_choice"].get("function")
                 if not specified_function:
                     raise ValueError(
                         "Incorrectly formatted `tool_choice`. Should be like "
                         "`{\"type\": \"function\","
                         " \"function\": {\"name\": \"my_function\"}}`")
```
Comment (DarkLight1337) [inline vllm/entrypoints/openai/protocol.py:486]: I think we should have different error message for this case.

Comment (shenoyvvarun) [reply vllm/entrypoints/openai/protocol.py]: Thanks for the comment. I fixed this in the new revision.

Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py]: This sounds like the user didn't specify a function, rather than the function not existing server side.

**Review by DarkLight1337 — APPROVED** [review]
Looks good now, thanks for the fix!

File: vllm/entrypoints/openai/protocol.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/entrypoints/openai/protocol.py]: Yes, it's more clear now.

**Pull request merged**